RFC: Templating `CARGO_TARGET_DIR` to make it the parent of all target directories (#3371)
rust-lang/rfcs: text/3371-cargo-target-base-dir.md

- Feature Name: `cargo_target_base_dir`
- Start Date: 2023-01-12
- RFC PR: [rust-lang/rfcs#3371](https://github.com/rust-lang/rfcs/pull/3371)
<!-- - Cargo Issue: [rust-lang/cargo#0000](https://github.com/rust-lang/cargo/issues/0000) -->

# Summary
[summary]: #summary

<!-- One paragraph explanation of the feature. -->

Introduce a new configuration option for cargo to tell it to move the workspace's target directory into a workspace-specific subdirectory of the configured absolute path, named `CARGO_TARGET_BASE_DIR`.

# Motivation
[motivation]: #motivation

<!-- Why are we doing this? What use cases does it support? What is the expected outcome? -->

The original motivating issue can be found here: [rust-lang/cargo#11156](https://github.com/rust-lang/cargo/issues/11156).

1. Not having to find and clean all `target/` dirs everywhere while not having all projects collide (which is the effect of setting `CARGO_TARGET_DIR` globally)

> 1. Being able to easily exclude a directory from saves (Apple's Time Machine, ZFS snapshots, BRTS, ...)

Reviewer: ```suggestion
1. Being able to easily exclude a directory from backups (Apple's Time Machine, ZFS and btrfs snapshots, ...)
```

## Impact on `cargo ...` calls

When calling `cargo` where `CARGO_TARGET_BASE_DIR` is active, `CARGO_TARGET_DIR` is set by all `cargo` calls that happen in a Cargo workspace, as much as possible. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool would have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today.

In the same vein, `cargo metadata` will fill the target directory information with the absolute path and make no mention of `CARGO_TARGET_BASE_DIR` since it can only be used in a single workspace at once.

### `cargo clean`

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for `CARGO_TARGET_BASE_DIR`: if it set, `cargo clean` will delete `CARGO_TARGET_BASE_DIR/<project-path>/` and not `target/`.

# Drawbacks

> [drawbacks]: #drawbacks

Author: Done

# Future possibilities
[future-possibilities]: #future-possibilities

<!--
Think about what the natural extension and evolution of your proposal would
be and how it would affect the language and project as a whole in a holistic
way. Try to use this section as a tool to more fully consider all possible
interactions with the project and language in your proposal.
Also consider how this all fits into the roadmap for the project
and of the relevant sub-team.

This is also a good place to "dump ideas", if they are out of scope for the
RFC you are writing but otherwise related.

If you have tried and cannot think of any future possibilities,
you may simply state that you cannot think of anything.

Note that having something written down in the future-possibilities section
is not a reason to accept the current or a future RFC; such notes should be
in the section on motivation or rationale in this or subsequent RFCs.
The section merely provides additional information.
-->

- Allowing relative paths: I feel this is counter-productive to the stated goal and have thought of no use for it, but it's entirely possible someone else will.
- Introduce remapping into the concept in some way.

> ## Use `CARGO_TARGET_BASE_DIR` as the default instead of `target`

Author: done

- Through the command line flag: `cargo build --target-base-dir /absolute/path/to/target/directories`

The given path must be absolute: setting `CARGO_TARGET_BASE_DIR` to an empty or relative path is an error (when used and not instantly overriden by `CARGO_TARGET_DIR`).

## Resolution order relative to `CARGO_TARGET_DIR`

The resolution order favors `CARGO_TARGET_DIR` in all its forms, in the interest of both backward compatibility and allowing overriding for a singular workspace:

`--target-dir` > `CARGO_TARGET_DIR` > `target-dir = ...` > `--target-base-dir` > `CARGO_TARGET_BASE_DIR` > `target-base-dir = ...`

See the final section for discussion on how to make `CARGO_TARGET_BASE_DIR` the default.

## Naming

In the example in the previous section, using `CARGO_TARGET_BASE_DIR` with `cargo build` produces named subdirectories. The name of those is partially deterministic: it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.

> This naming scheme is chosen to be simple for people to navigate but is **not considered stable**: the hashing method (and so the hash) will probably not change often but `cargo` offers no guarantee and may change it in any release. Tools that needs to interact with `cargo`'s target directory should not rely on its value for more than a single invocation of them: they should instead query `cargo metadata` for the actual value each time they are invoked.

Reviewer 2: > The name of those is partially deterministic: it is the name of the parent directory of the workspace's `Cargo.toml` manifest and an unspecified hash of the absolute path to the workspace's root, so building `work-project/crate-1` will still use the `/cargo-caches/work-project-<hash>/debug/...` directory for a `cargo build` call.

and

> This naming scheme ... is **not considered stable**:

Are contradictory.

Could we under specify his further and just use "work-project-<hash>` as an example of what could be done?  The most important part is we hash the manifest path.

This will let us avoid feeling the need to hash out these implementation details within the RFC and allow us to better explore what, between this and "cargo script", what approach we should take.

Author: Done, good point

## Handle possibly thousands of directories in a single `CARGO_TARGET_BASE_DIR` path

While a single dev machine is unlikely to have enough projects that the simple scheme of `<project name>-<hash>` will produce enough directories to slow down working in `$CARGO_TARGET_BASE_DIR/`, it could still happen, and notably in private CI, which are often less compartimentalized than public ones. Simple cruft over time (i.e, never calling `cargo clean` over years) could also make it happen, if much slower.

A quick design that would probably work and still be relatively navigable by humans is: instead of `project-name-<hash>/`, use `p/r/o/j/e/ct-name-<hash>/`. Here I used an arbitrary limit of 5 subdirectories but really, anything could be used. A problem occurs if a `.` is present in the directory name,but in that case we could simply append it to the symbol before like so: `proj....name/` -> `p/r/o/j..../n/ame-<hash>/`.

If the project name contains a space in the first 5 symbols, maybe replace it like so: `proj name/` -> `p/r/o/j-space/n/ame-<hash>/`.

I don't know of any filesystem that has troubles with `_` or `-` (often used as delimiters) so those have no special handling.

> If, on the other, we don't care (or care less) for human usability of the directories inside `CARGO_TARGET_BASE_DIR`, we could simply invert the format to use `<hash>-<project name>`: since the hash would have fixed length and set of characters, it would be easy to split into subdirectories without special casing and a simple `find . -n 'project name'` would point to the final target directory for the project (or all of them, if several similarly named projects exists on the machine).

Author: @estebank Here I discuss how to handle scaling.

I like the final idea I proposed but I'm waiting for other opinions before changing this RFC-wide

Reviewer 2: Something that will help explode the number of directories is ["cargo script"](https://github.com/rust-lang/cargo/issues/12207).  Originally, my plan was to use a shared CARGO_TARGET_DIR as a build-time optimization but with some recent discussions on per-user caching, I'm hopeful we can instead have per-script CARGO_TARGET_DIR. 

The current implementation is sort-of something like an unconditional `CARGO_TARGET_BASE_DIR=$CARGO_HOME/target` where the intermediate directories are `{hash[0:2]}/hash[2:]}` where the hash is just a "short hash".  Sharding the directory by 255 should be sufficient for any system....

Reviewer 2: - https://github.com/rust-lang/cargo/blob/master/src/cargo/core/workspace.rs#L387
- https://github.com/rust-lang/cargo/blob/master/src/cargo/util/hex.rs#L29

Author: underspecified naming scheme to allow for this

### `cargo clean`

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for `CARGO_TARGET_BASE_DIR`: if it set, `cargo clean` will delete `CARGO_TARGET_BASE_DIR/<project-path>/` and not `target/`.

# Drawbacks
[drawbacks]: #drawbacks

<!-- Why should we *not* do this? -->

## One more option to find the target directory

This introduces one more option to look at to find the target directory, which may complicate the life of external tools.

This is mitigated by having `CARGO_TARGET_DIR` entirely override `CARGO_TARGET_BASE_DIR`, so an external tool can set it and go on its way. Also, having `cargo` set `CARGO_TARGET_DIR` when inside a workspace where `CARGO_TARGET_BASE_DIR` is used will help current tools (those not yet using `cargo metadata`) continue working without trouble.

## Hitting windows path length limits

Depending on what naming scheme is used (e.g., a very long hash), we could hit the Windows path length limits if not careful.

> A mitigation for this is recommending a short prefix (in `CARGO_TARGET_BASE_DIR`) and using a has that doesn't include that many characters but those are only mitigations and do not fully fix the underlying problem.

Reviewer 3: ```suggestion
A mitigation for this is recommending a short prefix (in `CARGO_TARGET_BASE_DIR`) and using a hash that doesn't include that many characters but those are only mitigations and do not fully fix the underlying problem.
```

*Note: I looked at Bazel 5.4.0, the latest stable version as of this writing, things may change in the future or be different for older versions.*

# Unresolved questions
[unresolved-questions]: #unresolved-questions

<!--
- What parts of the design do you expect to resolve through the RFC process before this gets merged?
- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
-->

- Do we want to differentiate according to users ? `bazel` is a generic build tool, whereas `cargo` is not, so maybe differentiating on users is not necessary for us ?

> # Future possibilities

Reviewer 2: Can you add one about GC?  Basically, just link out to https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/last-use.20tracking

Author: done, added the link

```

When calling `cargo metadata` in the `symlink-to-crate` path, the result contains `"manifest_path": "/Users/poliorcetics/projects/actual-crate/Cargo.toml"` and `"workspace_root":"/Users/poliorcetics/projects/actual-crate"`. This behaviour means that symlinks won't change the final directory used inside `CARGO_TARGET_BASE_DIR`, or in other words: symbolic links are resolved.

### Handle possibly thousands of directories in a single `CARGO_TARGET_BASE_DIR` path

While a single dev machine is unlikely to have enough projects that the scheme of `<project-path>` will produce enough directories to slow down working in `$CARGO_TARGET_BASE_DIR/`, it could still happen, and notably in private CI, which are often less compartimentalized than public ones. Simple cruft over time (i.e, never calling `cargo clean` over years) could also make it happen, if much slower.

A proposed solution for `cargo` to split the hash into something like `$CARG_TARGET_BASE_DIR/hash[:2]/hash[2:4]/hash[4:]/...`. Since the naming scheme is considered an implementation detail, if this prove insufficient it could be change in a subsequent version of `cargo`.

## Impact on `cargo ...` calls

> When calling `cargo` where `CARGO_TARGET_BASE_DIR` is active, `CARGO_TARGET_DIR` is set by all `cargo` calls that happen in a Cargo workspace, as much as possible. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool would have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today.

Reviewer 4: I'm still not quite clear what it means exactly. Is it setting the `CARGO_TARGET_DIR` environment variable for build scripts? Proc macros? rustc? If so, why would it do that? None of those should be reading the CARGO_TARGET_DIR environment variable. When calling external `cargo-*` extensions, cargo doesn't know anything about manifests so it wouldn't be able to set the correct path. 

I'm not quite following the example for `cargo install` since it doesn't seem to be referring to an environment variable.

> When calling `cargo` where `CARGO_TARGET_BASE_DIR` is active, `CARGO_TARGET_DIR` is set by all `cargo` calls that happen in a Cargo workspace, as much as possible. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool would have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today.

Author: Fixed again

### `cargo clean`

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for `CARGO_TARGET_BASE_DIR`: if it set, `cargo clean` will delete `CARGO_TARGET_BASE_DIR/<project-path>/` and not `target/`.

### Providing backlinks

`targo` provides backlink (it links from it's own target directory to `<workspace>/target`) as a way for existing tools to continue working despite there being no `CARGO_TARGET_DIR` set for them to find the real target directory.

`cargo` does not for `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).

> When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a backlink by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.

Reviewer 5: Not sure what to do about non-NTFS but neither am I sure how well cargo works on non-NTFS (iirc there have been issues in the past and I know rustup requires NTFS). In any case I'd be minded to try a symlink first then fallback to a junction point. In the best case this has a good cross-platform story.

Though I'm not totally clear on the motivation for the symlink. Is it just a convenience feature? Or is it essential?

> When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a backlink by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.

Reviewer: It isn't absolutely strictly essential, but as the author of targo, I personally would not use this feature if it didn't create symlinks or some equivalent. As mentioned earlier, I'm fine with the state of affairs being that targo is a wrapper over this that also creates symlinks.

> When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a backlink by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.

Reviewer 5: I mean if creating a symlink fails for some reason, would it be ok to treat that as a warning rather than an error?

### `cargo clean`

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for the templated version: if it set, `cargo clean` deletes `/path/to/<manifest-path-hash>/` and not `target/`.

### Providing forward links

[`targo`][tg] provides forward link (it links from `<workspace>/target` to its own target directory) as a way for existing tools to continue working despite there being no explicit `CARGO_TARGET_DIR` set for them to find the real target directory.

`cargo` does not provide them for regular (untemplated) `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).

> When a templated `CARGO_TARGET_DIR` is used (in any form) (and not overriden, for example a templated env var overriden by a precise `--target-dir` option), it *will* use a forward link by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if the templated `CARGO_TARGET_DIR` wasn't set at all.

Reviewer 2: Not thrilled with having a heuristic for when to read `link-target-dir`.  Maybe there is context I'm missing about this since `link-target-dir` is under specified (see also https://github.com/rust-lang/rfcs/pull/3371#discussion_r1493516351).

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/07f899b4982fb5ced22a05172365a513c77f29b0

This has a two big advantages: not breaking external tools and giving easy access to artifacts produced by `cargo build/test/doc` to users (they're in the habit of typing `./target/debug/my-bin`, this would continue working with forward links).

A config option (CLI, `config.toml` and env var), `link-target-dir`, will be introduced to deactivate this behaviour, but it will `true` by default, for the reasons provided in favor of forward links just above.

#### Detailed working of forward links

> When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.

Reviewer 2: > a warning will be emitted but this will not prevent the rest of the action to go on

This could be very noisy for Windows users. I guess its reduced if we tell users that they can turn off forward links but I think I'd rather we fail silently.  Otherwise, people are likely to put this in their project or user config.  In the project config, it  now applies to everyone, regardless of platforms.  In the user config, this makes it harder to share a config across platforms.

Author: We could also make it only happen when the actual target directory is created and not after that, that way it will warn once then won't as long as there is no "cargo clean"

Maybe that's still too verbose and we would only warn in `--verbose` mode ?

Reviewer 2: > We could also make it only happen when the actual target directory is created and not after that, that way it will warn once then won't as long as there is no "cargo clean"

That might be sufficient.    We should probably put this down as an Unresolved Question for being determined during stabilization
- warning vs note
- always, on target creation, or with `--verbose`

Author: When to create/modify it is also more complicated, as you noted in https://github.com/rust-lang/rfcs/pull/3371#discussion_r1493519124, I'll move that to everything related to `Unresolved Questions` with the difficulties explained

Reviewer 2: I feel like the "only create forward link on target dir creation is an important note to call out (with the reason)

Author: I expand on forward links in 6540d11b8e2a2f4def265989a925932843ced8f3, it should hopefully be clearer

### Providing backlinks

Backlinks are metadata in templated `CARGO_TARGET_DIR` that links target directories back to the workspace they came from.

[`targo`][tg] uses them in its own form of the feature and `cargo` uses them too for templated `CARGO_TARGET_DIR`.

> While details of the stored data are left to the implementation (there is no need for `cargo` to expose this data directly, though it could be exposed through `cargo metadata` in the future, see the relevant section below), one could imagine using it to clean target directories whose corresponding workspace does not exist anymore when calling something like `cargo clean --all-workspaces` (doing it automatically is not possible, else any workspace on external disks would have its target directory cleaned up each time the disk is unmounted, which is way too aggressive a default).

Reviewer 2: More likely we'd track a mapping of workspaces to target dirs in our sqlite db used for garbage collection, see rust-lang/cargo#13136

> # Drawbacks

Reviewer 2: For `--target-dir`, some shells use `{}` and this would need escaping or quoting to be used on the command line

Author: I can't find a single mainstream shell where that's the case baring those that also interpret python ?

- Bash, Fish & ZSH use `${}`, not just `{}`
- Nushell uses `$"()"`

Reviewer 2: Bash has [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html).

However, it looks like if a comma or a sequence is not present, then it will be treated as a literal. For with a cargo-cult familiarity of brace expansion, like me, we should call it out and that so long as a variable doesn't have commas or `..`, we should not run into problems with brace expansion.

Author: Oh I completely forgot about that, you're right.

Quick tests on macOS with Bash and ZSH show that simply wrapping in `""` also works so it should be fairly easy to work around.

```bash
cd /tmp
mkdir tests

ls ./{tests,}   # SUCCESS

ls ./{tests}    # FAIL
ls "./{tests}"  # FAIL
ls "./{tests,}" # FAIL
```

Done in https://github.com/rust-lang/rfcs/pull/3371/commits/025b8effa206b047a18fceca2661db1ec4c4d554

[drawbacks]: #drawbacks

<!-- Why should we *not* do this? -->

## One more option to find the target directory

This introduces one more option to look at to find the target directory, which may complicate the life of external tools.

This is mitigated by the forward link provided by default by `cargo` when using the templated form of `CARGO_TARGET_DIR`.

## Hitting windows path length limits

Depending on what naming scheme is used (e.g., a very long hash), we could hit the Windows path length limits if not careful.

A mitigation for this is recommending a short prefix (in `CARGO_TARGET_DIR`) and using a hash that doesn't include that many characters but those are only mitigations and do not fully fix the underlying problem.

## Forward links

There a few cases where a symlink instead of a real dir will break programs: at least SQLite 3 can be configured to raise an error if the database is behind a symlink anywhere in its opening path, it's probably other programs can also be configured to check this (or do it by default). Since templated `CARGO_TARGET_DIR` won't become a default in this RFC, we are not breaking any existing use cases.

## Transition period

> During the transition period, any `CARGO_TARGET_DIR` that was defined as containing `{manifest-path-hash}` will change meaning. `cargo`, for at least one stable version of Rust, should provide warnings about this and point to either this RFC or its documentation to explain why the incompatiblity arised and how to fix it. In practice, paths with `{` or `}` in it are unlikely, even more with the exact key used by cargo here, so maybe no one will ever see the warning, but it's better than silently breaking workflows.

Reviewer 2: We have competing demands
- Users will want to set this and not worry about their Rust version
- Versions of Rust prior up to this being implemented will do a very wrong thing with variables set

We likely need to error, instead of warn.  While this will be frustrating for users who have to work with versions that error, this will raise awareness so they are less likely to set this with versions from before anything was implemented.

Reviewer 2: What if we instead treat the target dir as unset?  I'm unsure if that would be a positive or negative type of clever.

Author: I'm not in favor of this: silently treating the directory as unset would be surprising when one was explicitely (if wrongly) set. I regularly use this feature and I would prefer and error telling what to fix than my artifacts jumping places in some cases but not other

Reviewer 2: The "what to fix" is to turn off the feature completely when running specific versions of Rust (while you want it for others).

I feel like this is a "no win" situation and we need to make sure we call out the alternatives on this.

Author: Done in I think https://github.com/rust-lang/rfcs/commit/a821d25a624e1ddf3143e854a3f77dc9cff07dd9

<!--
Think about what the natural extension and evolution of your proposal would
be and how it would affect the language and project as a whole in a holistic
way. Try to use this section as a tool to more fully consider all possible
interactions with the project and language in your proposal.
Also consider how this all fits into the roadmap for the project
and of the relevant sub-team.

This is also a good place to "dump ideas", if they are out of scope for the
RFC you are writing but otherwise related.

If you have tried and cannot think of any future possibilities,
you may simply state that you cannot think of anything.

Note that having something written down in the future-possibilities section
is not a reason to accept the current or a future RFC; such notes should be
in the section on motivation or rationale in this or subsequent RFCs.
The section merely provides additional information.
-->

- Introduce remapping into the concept in some way.
- Introduce a form of garbage collection. Expanded upon this [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/last-use.20tracking).

## Adding `XDG_CACHE_HOME`, `HOME` or `CARGO_HOME` as templates

> Not all OSes use the XDG convention, notably Windows and macOS (though the latter can be somewhat made to) and it is very easy to define `CARGO_TARGET_DIR=${XDG_CACHE_HOME:-~/.cache}/cargo-target-directories/{manifest-path-hash}` if wanted by users.

Reviewer 2: This will be dealt with as part of rust-lang/cargo#1734

Author: I don't see what's relevant in that issue to the line here ? It's simply talking about adding some more templates

Reviewer 2: This is dismissive of providing this option.  Instead I would say that this is a desired solution but blocked on  rust-lang/cargo#1734 for defining the semantics / naming

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/b81b3edea0353fcf63568eac84a5e95360154459

It won't work in the `config.toml` but it will work with the environment variable and the command line option, both of which override the TOML config.

> It is certainly possible to add at least `{home}` and `{cargo_home}` but it can be done in the future without interfering at all with `{manifest-path-hash}`, making it relatively easy to defer and add if there is demand.

Reviewer 2: We could also have a variable for "user target base directory" (ie where cargo script writes to)

Author: As a template for the target directory ? I don't understand.

If you're talking about where `cargo script` will have its own target directory to write binaries and such, `{manifest-path-hash}` is already enough: it targets the manifest and not the directory to handle scripts that are in the same directory, so they will have their own target directory

Reviewer 2: We will have a canonical directory within `CARGO_HOME` (and wherever we move it when we do  rust-lang/cargo#1734) for target directories.  It would make sense to provide a variable for that base directory in the future.

Author: Done in https://github.com/rust-lang/rfcs/pull/3371/commits/b81b3edea0353fcf63568eac84a5e95360154459

Reviewer 2: I'm not seeing how that commit addresses this.

Author: Misunderstanding, better done in https://github.com/rust-lang/rfcs/commit/5dcfb2906a337605b6c25aa7a9d743462828b6a5

## Use templated `CARGO_TARGET_DIR` as the default instead of `target`

This option has several complications I'm not sure how to resolve:

1. How do we decide on good platform defaults ?
  - Subsequently, when platform defaults are decided, how do we ensure a new platform has a good default too ?

>   - `CARGO_HOME` is already criticized for being both a *cache* and a *config* home (using the XDG spec semantics), adding more local cache to it in the form of `CARGO_HOME/target-base-dir/` would not improve the situation and should probably not be done, but, if no good alternatives are found, there is precedent to use it for this.

Reviewer 2: These look to be more of problems being solved by rust-lang/cargo#1734

Reviewer 2: More my point is that this is a conversation separate from this RFC and should be removed, maybe with a foot note saying any concerns about this should see that issue.

Author: Done in e2b877084067d106b0237aa7e9fd3ba210bc4a68

### We really want to do this, how do we do it ?

Well, first, advertising of the option and its behaviour, as well as the forward link behaviour (so people know we're not just breaking tools for fun). After that, it becomes necessary to test it quite heavily to really ensure nothing has broken irremediably.

1. Introduce it as the default behaviour in nightly, wait for a few stable releases so that beta has it too and it can start being used for a few months after that so that esoteric setups can also try it.
2. Write a post saying "we'll do the change in version X" (where current version is like X-2 to warn 3 months before at least ?) and then only apply the change to projects where there is no untemplated `CARGO_TARGET_DIR` config set.

### A user still wants the current behaviour and not this RFC's

- We could use a config option to entirely deactivate it
- `CARGO_TARGET_DIR=target` would still be available

> - We could add special behaviour like `CARGO_TARGET_BASE_DIR=""` meaning "use current" directory

Reviewer 2: ```suggestion
- We could add special behaviour like `CARGO_TARGET_DIR=""` meaning "use current" directory
```

The first two are probably enough, the third is a bandaid.

### Expose template metadata

`cargo` will use backlinks in an implementation-defined form to keep track in the templated `CARGO_TARGET_DIR` of the relation from a target directory to its source workspace.

> In the future, we could envisage letting external tools and users access this data in a well-defined form through `cargo metadata`.

Reviewer 2: This is a child under "Use templated `CARGO_TARGET_DIR` as the default instead of `target`" but not seeing how its related.

What metadata would we be exposing and why?  If we don't have an idea of what problem we're solving with this future possibility, it doesn't seem all that likely to happen

Author: > What metadata would we be exposing and why?

I have no idea, that's why it's a future possibilities. Someone *may* ask for metadata in the future but since I have no idea what would be needed, it's staying a vague future possibility with no plan to do it

Reviewer 2: imo we should limit future possibilities to user-driven needs.

Author: Removed in https://github.com/rust-lang/rfcs/pull/3371/commits/8d1ddc29d8d637d0e2599d5aa31169704c30f511

When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a forward link by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.

This has a two big advantages: not breaking external tools and giving easy access to artifacts produced by `cargo build/test/doc` to users (they're in the habit of typing `./target/debug/my-bin`, this would continue working with forward links).

A config option (CLI, `config.toml` and env var), `link-target-dir`, will be introduced to deactivate this behaviour, but it will `true` by default, for the reasons provided in favor of forward links just above.

#### Detailed working of forward links

> When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.

Reviewer 2: Wanted to call this out for reviewers that the proposal currently suggests best-effort forward-links

Reviewer 2: For myself, I'm a little uncomfortable with the best-effort approach.  It feels sloppy.

Maybe if `link-target-dir` values were `auto` and `off`?

Author: I assume 'auto' would still be 'best effort' ?

Author: Or would it fail the command if the link cannot be created ?

Reviewer 2: The values for `link-target-dir` are currently unspecified in the RFC.  I am making the assumption that they are `true` and `false`.  My concern is with `true` being sloppy.  I am suggesting that we have an `auto` value that is a way out to explicitly recognize that it is "sloppy".

So with the framing of having an `auto` I could see us using
- `true`, `false`, `"auto"` (bool with a special string value)
- `always`, `auto`, `never` (like colors)

If its `true`  or `always`, it would error.

> When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.

Reviewer 2: From the motivation

> Allows easily having separate directories for Rust-Analyzer and Cargo itself

What happens if a `target/` dir or symlink already exists?  This is important both for transition and for the rust-analyzer use case.

Author: 1. We do nothing: low initial surprise, potential for bad behaviour later when the user tries to run binaries that are out of date
  - We could print a clear warning then to make it visible the cache is out of date ?
3. We delete and replace with a symlink: higher initial surprise, more compatibility with existing programs and tools probably ?

Reviewer 6: Having target forward link is a feature commonly used in package build systems like FreeBSD ports, MacPorts, etc.

Reviewer 2: If we don't update the symlink, then if someone updates their coonfig, it will be stale

If we do update it, it will break workflows where people set a per-tool target dir (CLI, rust-analyzer, etc).

Almost seems like we'd want to specialize config file (fairly static) with env variables / `--target-dir` (more dynamic) but even that could likely cause problems and be even more confusing because nothing else in cargo acts like this.

Reviewer 6: Usually package systems I mentioned use following: 

* whole path is used as a subfolder aster prefix, not hash. E.g. for a project in /path/to/projest build folder is created under /prefix/path/to/folder. 
* These folders and symlink are created when project is built (pre-build step, if symlink or folder doesn't exists)
* symlinks are used by build systems and all tools. This ease algorithms used in system itself, tools and make behaviour backward compatible if anything changes. 
* symlinks and build folders are removed on clean
* if a user just removed symlink and moved project folder, he's responsible for a cleanup.

Reviewer 2: @eirnym does that prior art deal with the project being concurrently built with different target directories?

Reviewer 2: What we likely should do is if people have background builders where they don't care about the artifacts, tell them to set `target-dir-link = `false`.

Reviewer 6: @epage I agree, that on Windows it would be a problem, I use mostly macOS, Linux and FreeBSD, so I mentioned only best systems I've seen so far. Hashes would be great if a user can cleanup caches for an exact project without just removing everything because it's impossible to find it by just looking on it.

> does that prior art deal with the project being concurrently built with different target directories?

I'm not fully understand the meaning, please correct me if I guessed wrong.

1. If I understand this question as "does package system build concurrently after creating actual build directories?" yes, all of these systems do build concurrently if possible. 
2. If I understand this question as "we have different CPU/build targets, how we would create a folder structure for them?", as far as I understood this RFC, that only base `/path/to/project/target` folder would be moved, and the contents would stay the same, as done for example by Xcode and by cargo (at the moment): they create base folder and create subfolders as needed. This system works already and I don't see a reason to change it, except of the actual placement for project/_target_ folder

Reviewer 6: Give me a shout if you need documentation references for these systems

Author: Done in fac2d9c0852b8b043d3baf22f3fe5313de743072

Reviewer 2: Can we make sure we also cover the prior art?

Author: I assume prior arts for forward links ? barring `targo`, I have not found any (I suspect I just don't know the domain terms used by other build systems)

- `bazel`: nothing I could find, instead they document their naming scheme for the build paths, allowing users to very easily build their own forward link if wanted
- `go build`: builds a single binary directly in the root (or cwd)
- `make`, `configure`, `cmake`: everyone does it differently but the convention is often `<project root>/build/` dir or sometimes a `dist/` one

I haven't reviewed every build system in existence, I could look toward the JS ecosystem, they probably have one package manager with forward link but I have never worked with JS so I wouldn't even know where to start

Reviewer: Yeah, I'm not aware of any prior art for forward links for the target directory itself. I believe Buck2 creates symlinks (or maybe hard links) for executables to a content-addressed store.

Reviewer 2: Was referring to the Ports system talked about in @eirnym's posts.

Author: I would be interested in links to docs for those then, because I know nothing about them and build tools are not something I'm willing to spend time diving in to find that docs don't even exist in the end.

Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for `CARGO_TARGET_BASE_DIR`: if it set, `cargo clean` will delete `CARGO_TARGET_BASE_DIR/<project-path>/` and not `target/`.

### Providing backlinks

`targo` provides backlink (it links from it's own target directory to `<workspace>/target`) as a way for existing tools to continue working despite there being no `CARGO_TARGET_DIR` set for them to find the real target directory.

> `cargo` does not for `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).

Reviewer 2: > `cargo` does not for `CARGO_TARGET_DIR`

typo?

> When `CARGO_TARGET_BASE_DIR` is used (in any form) and not superseded by other configurations (`CARGO_TARGET_DIR`), it *will* use a backlink by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if `CARGO_TARGET_BASE_DIR` and `CARGO_TARGET_DIR` weren't set at all.

Reviewer 2: I think there might some terminology confusion.  I thought a "backlink" was metadata inside the actual target directory pointing back to the project.  This allows you to walk all of the centralized target directories and find out if the associated project still exists.

Reviewer: Yes, when I say backlinks I mean metadata stores in the targo directory that lists out the set of Rust project directories. The forward link is from workspace-dir/target to the central store.
